switch to using CSTRMarshaller instead of AnsiBSTR

mangod9 opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/54820. 

cc @dotnet/crossgen-contrib  @MichalStrehovsky

mangod9 committed:
switch to using CSTRMarshaller instead of AnsiBSTR

mangod9 committed:
cleanup

mangod9 commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs:
```diff
@@ -1581,13 +1581,7 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
 
     class AnsiStringMarshaller : Marshaller
     {
-        internal override bool CleanupRequired
```
tried to mimic stub for `ILCSTRMarshaler` from ilmarshallers.cpp, where its marked with no CleanupRequired.

jkotas replied:
That does not sound right. This needs cleanup to free the buffer.

mangod9 replied:
so looking further at `ILCSTRMarshaler` native cleanup is [required](https://github.com/dotnet/runtime/blob/317783916d5695d86b087b45220a28c5d8db9fc6/src/coreclr/vm/ilmarshalers.cpp#L477), but not for managed. Doesnt look like this cg2 implementation makes such a differentiation. Perhaps that is the root of the problem since the failure only occurs during native->Managed string marshalling?

mangod9 replied:
I have updated the `CleanupManaged` logic with the one from which seems to work https://github.com/dotnet/runtime/blob/317783916d5695d86b087b45220a28c5d8db9fc6/src/coreclr/vm/ilmarshalers.cpp#L483.

mangod9 commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs:
```diff
@@ -1603,24 +1597,76 @@ protected override void TransformManagedToNative(ILCodeStream codeStream)
             // ANSI marshalling. Allocate a byte array, copy characters
             //
 
-#if READYTORUN
```
do we have a usecase for non `READYTORUN`, or should we cleanup the codebase?

jkotas replied:
These READYTORUN ifdefs is allows easier sharing of the code with the NativeAOT branch.

mangod9 replied:
ok, thanks. Have keep `!READYTORUN` as-is then.

mangod9 commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs:
```diff
@@ -1603,24 +1597,76 @@ protected override void TransformManagedToNative(ILCodeStream codeStream)
+
+            bool bPassByValueInOnly = In && !Out && !IsManagedByRef;
+            ILLocalVariable localBuffer = emitter.NewLocal(Context.GetWellKnownType(WellKnownType.IntPtr));
+            if (bPassByValueInOnly)
```
is there a reason why this logic cant move to `System.StubHelpers.CSTRMarshaler` ?

jkotas replied:
Yes, it would be nice to have majority of the logic in C#, and minimize amount of the hand-emitted IL.

mangod9 commented:
@AaronRobinsonMSFT as well.. perhaps we should just fix the `out string` case and refactor these helpers in 7.

AaronRobinsonMSFT commented:
/cc @jkoritzinsky @elinor-fung

mangod9 committed:
Add CleanupManaged back

Updated the cleanup logic to the one implemented in ILOptimizedAllocMarshaler::EmitClearNative

jkotas commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs:
```diff
@@ -1581,6 +1581,10 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
     {
+        const int MAX_LOCAL_BUFFER_LENGTH = 256; // TODO: Is this accurate on all platforms?
+
+        private ILLocalVariable m_localBuffer = default;
```
```suggestion
        private ILLocalVariable _localBuffer = default;
```
Our convention for C# is to use `_` prefix for instance fields.

mangod9 replied:
yup will rename, once current CI job completes.

mangod9 committed:
CR feedback

mangod9 commented:
crossgen2 outerloop failures are comparable to main runs.

mangod9 commented:
@VSadov since we saw this show up as part of this issue: https://github.com/dotnet/runtime/issues/54759. Any comments / concerns with this change? 

Wonder how we hadn't seen any previous issues here since libraries have been compiled using cg2 for a few previews now and a simple interop call with `out string` should hit this case.

VSadov commented:
@mangod9 - the fix look reasonable and the issue is very likely the same as one observed in #54759. 

As for why we have not seen this sooner - maybe seeing a combination of "out" parameters in interop scenario, on Linux, with crossgen2 just took time to observe...

mangod9 commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs:
```diff
@@ -1581,6 +1581,10 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
 
     class AnsiStringMarshaller : Marshaller
     {
+        const int MAX_LOCAL_BUFFER_LENGTH = 256; // TODO: Is this accurate on all platforms?
```
MAX_PATH on windows is `260` so will update this. assume same value is used on other platforms in the PAL.

mangod9 committed:
update MAX_LOCAL_BUFFER_LENGTH

MichalStrehovsky commented:
Do we need a regression test?

mangod9 commented:
yeah I am hoping there are some interop tests already covering `out string` scenario which we could add to the r2r pipeline. @AaronRobinsonMSFT @jkoritzinsky @elinor-fung do you happen to know if any exist?

jkotas commented:
We have a lot of tests for this. The problem is that the interop stub generation in crossgen is currently only enabled for CoreLib, so it is impossible to test.

We would need to enable it for everything in CoreLib version bubble to make it testable by changing [this](https://github.com/dotnet/runtime/blob/333c4e72b83a08288db929ac56c0e336c936b655/src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilationModuleGroupBase.cs#L362) to `if (_versionBubbleModuleSet.Contains(method.Context.SystemModule))`.

mangod9 commented:
which is being done in this PR: https://github.com/dotnet/runtime/pull/54847? So a combination of the two PRs (without skipping `AnsiStringMarshaller`) would suffice?

jkotas commented:
Yes, assuming we have appropriate large version bubble test leg.

mangod9 committed:
generate stubs for all methods if corelib is within the bubble.

mangod9 commented:
with the latest change I am hitting an issue similar to this: https://github.com/dotnet/runtime/issues/51189. so more of these marshallers would need updates to work with R2R

jkotas commented:
I would recommend undoing the last change, and deal with the testing later.

We need a unified intentionally designed set of helpers for marshalling strings that are shared between runtime-generated, crossgen-generated and source-generated interop.

mangod9 commented:
> I would recommend undoing the last change, and deal with the testing later.

yeah that seems like the best approach to get this merged. Will revert that change.

mangod9 committed:
Revert "generate stubs for all methods if corelib is within the bubble."

This reverts commit 8575306e73ee6b009bc05a8153abd7afa4380dfc.

mangod9 commented:
ok, CI is good now. crossgen2-composite failures are at par with main.

MichalStrehovsky commented:
Btw, disabling the Utf8Marshaller would be as simple as deleting these two lines: https://github.com/dotnet/runtime/blob/d431d6a15726797965336176efe9ae15aba241fd/src/coreclr/tools/aot/ILCompiler.ReadyToRun/Interop/IL/Marshaller.ReadyToRun.cs#L28-L29. I don't think Utf8Marshaller as it is right now works with crossgen2 at all.

Looks like all of this slipped through the cracks because nobody realized it's pretty much not exercised by anything.

mangod9 commented:
yeah will do that as a separate change. Believe there are other ones (like `UnicodeString`) which need to excluded for R2R.

mangod9 commented:
Any other concerns with this?

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs:
```diff
@@ -1581,6 +1581,10 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
 
     class AnsiStringMarshaller : Marshaller
     {
+        const int MAX_LOCAL_BUFFER_LENGTH = 260 + 1; // MAX_PATH + 1 
+
+        private ILLocalVariable _localBuffer = default;
```
I think unused locals would break the build.

```suggestion
#if READYTORUN
        const int MAX_LOCAL_BUFFER_LENGTH = 260 + 1; // MAX_PATH + 1 

        private ILLocalVariable _localBuffer;
#endif
```

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs:
```diff
@@ -1605,12 +1609,75 @@ protected override void TransformManagedToNative(ILCodeStream codeStream)
+                // String.Length + 2
+                LoadManagedValue(codeStream);
+                var stringLen =
+                            Context.SystemModule.GetKnownType("System", "String")
```
```suggestion
                            Context.GetWellKnownType(WellKnownType.String)
```

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs:
```diff
@@ -1605,12 +1609,75 @@ protected override void TransformManagedToNative(ILCodeStream codeStream)
             codeStream.EmitLdc(flags);
             LoadManagedValue(codeStream);
+
+            if (_localBuffer != default)
```
Might be better to use `ILLocalVariable?` as the type of _localBuffer. This condition will also be true if we created a local and it was the first local (index 0).

mangod9 replied:
makes sense, changed in latest commit.

MichalStrehovsky reviewed (approved):
Looks good otherwise.

mangod9 committed:
CR feedback

mangod9 committed:
use wellknowntype
